### ISSUE TITLE
Docker guidance

### ISSUE THREAD

**Issue by JordanMarr** [AngelMunoz/Migrondi#34]
This is probably out of scope, but do you have any guidance for running Mirondi automatically when starting a docker container?

I'd like to have a more unified way of initializing the SqlHydra databases with data. Not only for initializing the adventureworks db, but some providers need custom tables added for provider specific features. For example, postgres needs to have a schema added to test enum support. It would be cool to have a standard scripts folder that would just work for each provider.

**Comment by JordanMarr** [comment]
I haven't actually written any code yet. (Still in the research phase.)

Each provider has its own separate folder with the files it needs. So, I think a separate migration folder within each provider folder would be a nice way to go. 

Do you support Oracle? I don't see it listed, but I'm not sure if you are actually doing any Provider specific stuff or not.

**Comment by AngelMunoz** [comment]
Ahh cool that sounds good, I think that could be done via an F# script/program that runs these programmatically. A more concise way (which is not built right now) would be to accept certain configuration values from the CLI in this case
```
migrondi up --driver sqlite --migrations ./sqlite
migrondi up --driver mysql --migrations ./mysql
```
Or something like that.

For the current programatic way to do that would be:

```fsharp
#r "nuget: Migrondi.Core, 1.0.0-beta-010"


open Migrondi.Core
open System.Threading.Tasks


let sqliteConfig =
    { MigrondiConfig.Default with
        driver = MigrondiDriver.Sqlite
        connection = "Data Source=database.db"
        migrations = "./sqlite" }

let postgresConfig =
    { MigrondiConfig.Default with
        driver = MigrondiDriver.Postgresql
        connection = "Host=localhost;Port=5432;Username=postgres;Password=postgres;Database=postgres"
        migrations = "./postgres" }

let mssqlConfig =
    { MigrondiConfig.Default with
        driver = MigrondiDriver.Mssql
        connection = "Server=localhost;Database=master;User Id=sa;Password=Password123;"
        migrations = "./mssql" }

let sqlite = Migrondi.MigrondiFactory(sqliteConfig, ".")

let postgres = Migrondi.MigrondiFactory(postgresConfig, ".")

let mssql = Migrondi.MigrondiFactory(mssqlConfig, ".")

// run them concurrently if you will or on your own demand
Task.WaitAll(
    task { do! sqlite.DryRunUpAsync() :> Task },
    task { do! postgres.DryRunUpAsync() :> Task },
    task { do! mssql.DryRunUpAsync() :> Task }
)
```

Which could then be run after installing dotnet in your dockerfile, to be honest I haven't tried any docker workflows so I may be just rambling too much here.

> Do you support Oracle? 

Currently I support what [RepoDB](https://repodb.net/tutorial/installation) supports because that's the backing mechanism I have to run queries, but to be honest I think I could make the effort to drop it and go at the ADO provider to do all I have to do as I don't really do any RepoDB specific stuff, I just run user scripts, insert and query out.


I plan to publish v1 at the end of april, so this is a great time to review this kind of feedback I appreciate it, and also let me know if that's somewhat the things you'd like to see in the library (in case you choose to use it) or if it is kind of out of place

**Comment by JordanMarr** [comment]
Just me thinking out loud:
My original intention for SqlHydra was to have the Docker scripts auto-run to create the sample AdventureWorks database that is used for testing for each provider. 
I did partially achieve this, but there are discrepancies between the way the different provider containers work that prevent some of them from fully loading, which results in a confusing post setup dilemma that I always forget how to do.

Issues include:
* Sql Server sometimes fails to apply .sql scripts
* Oracle, to save space, does not run out of the box. Instead, it just loads an installer that takes 30m to pull files from the cloud and then install an instance. (Had I known this, I would have never supported Oracle!)

Now it occurs to me that even if I add a migrations tool, launching it from Docker is likely to suffer from similar issues. 
Plus, it's hard to setup and time intensive to troubleshoot.

So, it now occurs to me that I could probably simplify everything by just run the migrations as part of the test suite.

Which brings me to my question:
How does Migrondi check to see if a script has been run (so that it doesn't run multiple times)?
I currently use EvolveDb which creates a metadata table in the database that stores each script name (along with a hash to detect if the script has been changed since it was applied, in which case it throws an error). 

Is this stamped directly into the script files by Migrondi?

**Comment by AngelMunoz** [comment]
It is a very similar process, it creates a table within the db and it inserts the record in the database once it has been run.

I don't do file/db checksums to see if contents have changed but that's something I'm considering given that I've heard that mentioned a couple times already (in un-related to this repo talks tough :laughing:)

> Oracle, to save space, does not run out of the box. Instead, it just loads an installer that takes 30m to pull files from the cloud and then install an instance. (Had I known this, I would have never supported Oracle!)

Yikes!

**Comment by JordanMarr** [comment]
With EvolveDb, if a script changes after it's been applied, then I have to run a "repair" command which just overwrites the checksum column.

I guess it's good because it makes the script change very explicit, so you are forced to sort of "approve" it by running repair.
The downside is that it's kind of annoying to have to go back and update the script.

But, it seems like a good safety feature to have if you want to assume that your migration scripts should be an "immutable", append-only style log (which seems congruent with F# in general).

**Comment by kaashyapan** [comment]
Hi there! I was evaluating the migration landscape and love the fact that this can be used in a script or as standalone tool.
I wrote the [sqlc plugin for F#](https://github.com/kaashyapan/sqlc-gen-fsharp) and have been using [Atlas go](https://atlasgo.io/atlas-vs-others) as a migration tool until they decided to screw by adding a dependency to their cloud service.

I just wanted to bring this to your attention based on @JordanMarr comment above to see if we can steal any good ideas from that project. Declarative stateless schema definitions rather than checksums. Maybe leverage [ThrowawayDb](https://github.com/Zaid-Ajaj/ThrowawayDb)

![Image](https://github.com/user-attachments/assets/0a70a668-fb14-4da1-8ecd-aaf2030b9fce)

**Comment by AngelMunoz** [comment]
@kaashyapan that sounds like an interesting approach to be honest, I would like to take some time to learn more about the declarative stateless schema definitions, I found https://supabase.com/docs/guides/local-development/declarative-database-schemas which kind of sets my mind on how this looks

My understanding though would be that generated migrations might look different from person to person depending on their own database schema state, right? 

Taking this as an example 
```
.
└── supabase/
    ├── schemas/
    │   ├── employees.sql
    │   └── managers.sql
    └── migrations/
        ├── 20241004112233_create_employees_table.sql
        ├── 20241005112233_add_employee_age.sql
        └── 20241006112233_add_managers_table.sql
```
I guess for these cases migrations should not be committed into git as the tool should provide the sql gen every time it is ran

Is it what you would expect or am I actually lost?

**Comment by kaashyapan** [comment]
> My understanding though would be that generated migrations might look different from person to person depending on their own database schema state, right?

That is correct. If each developer has a different starting state, it indicates schema drift. Even if the starting state is stale the final state will be the same.

Essentially the developer is defining the expected state. Whether that is achieved through a CREATE or an ALTER is not his concern. What resources need to be dropped to achieve the final state is not his concern.

The migrations can be committed to have a record of change history to see what was approved. Usually there is an approval process in the pipeline after the SQLs are generated from the diff (`supabase migration up`). Useful to see if any indices were dropped / troubleshoot performance regressions. The migrations would not be rerun by developers rebasing on production. They would regenerate their own diffs (ideally identical).
